File: my-folder/1073-number-of-enclaves/solution.cpp
class Solution {
public:

    void dfs(int i,int j,int rows,int cols,vector<vector<int>>& grid)
    {
        if (i>=0 && i<rows && j>=0 && j<cols and grid[i][j]==1)
        {
            grid[i][j] = 'T';
            dfs(i+1,j,rows,cols,grid);
            dfs(i-1,j,rows,cols,grid);
            dfs(i,j+1,rows,cols,grid);
            dfs(i,j-1,rows,cols,grid);
        }
        else
        {
            return;
        }
            
    }
    int numEnclaves(vector<vector<int>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();

        // set all boundary lands and connected land to 't'
        for (int i=0; i<rows;i++)
        {
            for (int j=0; j<cols;j++)
            {
                if (i==0 or i==rows-1 or j==0 or j==cols-1)
                {
                    dfs(i,j,rows,cols,grid);
                }
            }
        }

        // the remaining 1s are the region that can not be walked off
        int count = 0;
        for (int i=0; i<rows;i++)
        {
            for (int j=0; j<cols;j++)
            {
                if ( grid[i][j] ==1)
                    count+=1;   
            }
        }
        return count;

    }
};

File: my-folder/1073-number-of-enclaves/solution.py
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        visited = [ ['U']*n for _ in range(m)]
        ans = 0

        def dfs(i,j):
            if i < 0 or i >= m or j < 0 or j >= n: return
            if visited[i][j] != 'U': return
            visited[i][j] = 'V'
            for dx, dy in [(0,1), (0,-1), (1,0), (-1,0)]:
                new_x, new_y = i + dx  , j + dy
                if 0 <= new_x < m and 0 <= new_y < n and grid[new_x][new_y] == 1:
                    dfs(new_x, new_y)
            return
                    
        for i in range(m):
            for j in range(n):
                if i == 0 or j ==0 or i == m-1 or j == n-1:
                    if grid[i][j] ==1 and visited[i][j] == 'U':
                        dfs(i,j)
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] and visited[i][j] == 'U':
                    ans+=1
        return ans

